From 620b0ffba331f1d44b7dc7c4db62e5ea1a18d7a6 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Sun, 19 Oct 2014 17:14:10 +0200
Subject: [PATCH] video: sunxi-hdmi: Deduce VIC from timing when mode is taken
 from EDID.

Set VIC in AVI info frame in case the video mode is DISP_TV_MODE_EDID.
Also correct a possible array bounds violation in get_video_info().

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/sunxi/hdmi/hdmi_core.c |   28 +++++++++++++++++++++++++---
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/drivers/video/sunxi/hdmi/hdmi_core.c b/drivers/video/sunxi/hdmi/hdmi_core.c
index a6dd01f..ce2efed 100644
--- a/drivers/video/sunxi/hdmi/hdmi_core.c
+++ b/drivers/video/sunxi/hdmi/hdmi_core.c
@@ -217,9 +217,9 @@ __s32 Hpd_Check(void)
 
 __s32 get_video_info(__s32 vic)
 {
-	__s32 i, count;
-	count = sizeof(video_timing);
-	for (i = 0; i < count; i++)
+	__s32 i;
+
+	for (i = 0; i < ARRAY_SIZE(video_timing); i++)
 		if (vic == video_timing[i].VIC)
 			return i;
 
@@ -227,6 +227,25 @@ __s32 get_video_info(__s32 vic)
 	return -1;
 }
 
+__s32 vic_from_video_timing(const struct __disp_video_timing *mode)
+{
+	__s32 i;
+	struct __disp_video_timing temp_mode;
+	
+	memcpy(&temp_mode, mode, sizeof(temp_mode));
+
+	for (i = 0; i < ARRAY_SIZE(video_timing); i++)
+	{
+	      temp_mode.VIC = video_timing[i].VIC;
+	      
+	      if (memcmp(&temp_mode, &video_timing[i], sizeof(temp_mode)) == 0)
+			return video_timing[i].VIC;
+	}
+	
+	return HDMI_EDID;
+}
+
+
 static __s32 get_audio_info(__s32 sample_rate)
 {
 	__s32 vic_tab;
@@ -301,6 +320,9 @@ __s32 video_config(__s32 vic)
 		return 0;
 	else
 		video_mode = vic;
+	
+	if (vic == HDMI_EDID)
+		vic = vic_from_video_timing(&video_timing[video_timing_edid]);
 
 	if ((vic == HDMI1440_480I) || (vic == HDMI1440_576I))
 		dw = 1; /* Double Width */
-- 
1.7.9.5

